cpp/include/cudf/sorting.hpp

```diff
@@ -184,5 +184,63 @@ std::unique_ptr<column> rank(
   bool percentage,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
 
+/**
+ * @brief Performs a lexicographic segmented sort of a table
+ *
+ * If segment_offsets contains values larger than number of rows, behaviour is undefined.
+ * @throws cudf::logic_error if `values.num_rows() != keys.num_rows()`.
+ * @throws cudf::logic_error if `segment_offsets` is not `size_type` column.
+ *
+ * @param values The table to reorder
+ * @param keys The table that determines the ordering of elements in each segment
+ * @param segment_offsets The column of `size_type` type containing start offset index for each
+ * contiguous segment.
+ * @param column_order The desired order for each column in `keys`. Size must be
+ * equal to `keys.num_columns()` or empty. If empty, all columns are sorted in
+ * ascending order.
+ * @param null_precedence The desired order of a null element compared to other
+ * elements for each column in `keys`. Size must be equal to
+ * `keys.num_columns()` or empty. If empty, all columns will be sorted with
+ * `null_order::BEFORE`.
+ * @param mr Device memory resource to allocate any returned objects
+ * @return table with elements in each segment sorted.
+ *
+ */
+std::unique_ptr<table> segmented_sort(
+  table_view const& values,
+  table_view const& keys,
+  column_view const& segment_offsets,
+  std::vector<order> const& column_order         = {},
+  std::vector<null_order> const& null_precedence = {},
+  rmm::mr::device_memory_resource* mr            = rmm::mr::get_current_device_resource());
+
+/**
+ * @brief Performs a lexicographic segmented sort of the list in each row of a table of list columns
```

Comment:
I don't understand this function. How do you do a lexicographic sort of a `table` of list columns?

Reply:
Is this clear?
`Performs a lexicographic sort of each list in the table.`

Reply:
So is it the same as calling sort independently on each column?

Reply:
no. It will be similar to calling sort on each row of the table. Each row is a list.
The relative row order in the table remains same. In each row, the elements in list are sorted.

Reply:
So it's the same as calling `sort_lists` on each column in the table?

Reply:
no. it's similar to treating each row as table, then sort it. then repeat it for all rows.

Copying the example from #6541

| row# | a | b | c |
| - | - | - | - |
| 0 | [21, 22, 23, 22] | [13, 14, 12, 11] | ["a", "b", "c", "d"]
| 1 | [22, 21, 23, 22] | [14, 13, 12, 11] | ["a", "b", "c", "d"]

Here [...] is a list.


| row# | a | b | c |
| - | - | - | - |
| 0 | <ul><li>21</li><li>22</li><li>23</li><li>22</li></ul> | <ul><li>13</li><li>14</li><li>12</li><li>11</li></ul> | <ul><li>"a"</li><li>"b"</li><li>"c"</li><li>"d"</li></ul> |
| 1 | <ul><li>22</li><li>21</li><li>23</li><li>22</li></ul>| <ul><li>14</li><li>13</li><li>12</li><li>11</li></ul> | <ul><li>"a"</li><li>"b"</li><li>"c"</li><li>"d"</li></ul> |

`sort_lists(values={a,b,c}, keys={a,b})` output will be

| row# | a | b | c |
| - | - | - | - |
| 0 | <ul><li>21</li><li>22</li><li>22</li><li>23</li></ul> | <ul><li>13</li><li>11</li><li>14</li><li>12</li></ul> | <ul><li>"a"</li><li>"d"</li><li>"b"</li><li>"c"</li></ul> |
| 1 | <ul><li>21</li><li>22</li><li>22</li><li>23</li></ul>| <ul><li>13</li><li>11</li><li>14</li><li>12</li></ul> | <ul><li>"b"</li><li>"d"</li><li>"a"</li><li>"c"</li></ul> |

Reply:
Does that require all of the lists in a given row to have the same number of elements? What would the result be for this input?

| row# | a | b | c |
| - | - | - | - |
| 0 | [21, 22, 23, 22] | [13, 14] | ["a", "c", "d"]
| 1 | [22, 21, 23, 22] | [14] | ["c", "d"]

Reply:
Yes. 
For this input, it will throw a logic error.